source/loader/layers/sanitizer/asan_interceptor.hpp

```diff
@@ -0,0 +1,143 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file ur_sanitizer_layer.cpp
+ *
+ */
+
+#pragma once
+
+#include "common.hpp"
+
+#include <map>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+namespace ur_sanitizer_layer {
+
+enum USMMemoryType { DEVICE, SHARE, HOST, MEM_BUFFER };
+
+struct USMAllocInfo {
+    uptr AllocBegin;
+    uptr UserBegin;
+    uptr UserEnd;
+    size_t AllocSize;
+    USMMemoryType Type;
+};
+
+enum class DeviceType { UNKNOWN, CPU, GPU_PVC, GPU_DG2 };
+
+struct DeviceInfo {
+    DeviceType Type;
+    size_t Alignment;
+    uptr ShadowOffset;
+    uptr ShadowOffsetEnd;
+
+    // Lock InitPool & AllocInfos
+    ur_shared_mutex Mutex;
+    std::vector<std::shared_ptr<USMAllocInfo>> AllocInfos;
+};
+
+struct QueueInfo {
+    ur_mutex Mutex;
+    ur_event_handle_t LastEvent;
+};
+
+struct ContextInfo {
+
+    std::shared_ptr<DeviceInfo> getDeviceInfo(ur_device_handle_t Device) {
+        std::shared_lock<ur_shared_mutex> Guard(Mutex);
+        assert(DeviceMap.find(Device) != DeviceMap.end());
+        return DeviceMap[Device];
+    }
+
+    std::shared_ptr<QueueInfo> getQueueInfo(ur_queue_handle_t Queue) {
+        std::shared_lock<ur_shared_mutex> Guard(Mutex);
+        assert(QueueMap.find(Queue) != QueueMap.end());
+        return QueueMap[Queue];
+    }
+
+    std::shared_ptr<USMAllocInfo> getUSMAllocInfo(uptr Address) {
```

Comment:
this doesn't appear to be used anywhere.

Reply:
It will be used to implement the logic of memory releasing.